pkg/store/proxy.go

```diff
@@ -515,7 +515,8 @@ func storeMatches(s Client, mint, maxt int64, storeMatcher [][]storepb.LabelMatc
 		return false, nil
 	}
 	match, err := storeMatchMetadata(s, storeMatcher)
-	if err != nil || !match {
+	// Return result here if no matchers set.
+	if len(matchers) == 0 || err != nil || !match {
```

Comment:
Doesn't `labelSetsMatch()` already return `true, nil` if `matchers` is `nil`? In `labelSetMatches` we `range` over nil and because it's empty, we simply jump to the end and return `true, nil`, right?

Reply:
`return labelSetsMatch(s.LabelSets(), matchers)` If you go to `labelSetsMatch` here, we need to first evaluate `s.LabelSets()` https://github.com/thanos-io/thanos/blob/master/pkg/query/storeset.go#L277.

I want to avoid this because it needs to get the lock and alloc a string slice.